c10/cuda/CUDAGuard.h

```diff
@@ -76,7 +77,9 @@ struct CUDAGuard {
 /// CUDAGuard for when you can use this.
 struct OptionalCUDAGuard {
   /// Create an uninitialized OptionalCUDAGuard.
-  explicit OptionalCUDAGuard() : guard_() {}
+  explicit OptionalCUDAGuard() : guard_() {
+    c10::cuda::set_device(c10::cuda::current_device());
```

Comment:
I will add a detailed comment about this. In two words, the current device is not always set. 
Simple example:
```
import torch
x = torch.empty(1, device="cuda:1")
torch.cuda.current_device() # Will return 0
```
Although current device is 0, `cudaSetDevice(0)` was never called. It will be called in class instantiation here.

Reply:
Isn't the whole point here to make sure that these 3 lines never initialize device 0 ?
Do you have a full example of what will fail if you dont have this line?

Reply:
Please remember that this is a constructor of `OptionalCUDAGuard`. The target was avoid unnecessary `cudaSetDevice` at the destructors.
Simply put, if I do not have this, some of DDP functionality (where `OptionalCUDAGuard`s are widely used)  would simply hang: https://github.com/pytorch/pytorch/actions/runs/4482542370/jobs/7881218551 because device mismatches. 
The moment I introduced this line in 75496ce, there is no more hangs. https://github.com/pytorch/pytorch/actions/runs/4494166949/jobs/7906310163.
Despite, I have not seen failures with the stream guards, I believe it is must have too.